Add ESM and rollup build

hugo-vrijswijk opened the pull request:
First of all, thanks for this awesome library!

Fixes #94

Adds a rollup config file to build ESM, CJS, and UMD bundles. Based on the source in src/fuzzysort.js.

Checks in the new esm file and the updated cjs and umd files.

Also updates the package.json to include `module` and `browser` fields for the ESM and UMD bundles. This lets bundlers, node or CDN's know which version to use. 

I've tried to keep things fairly streamlined. Running 'test' automatically builds the files. The `src/fuzzysort.js` and `fuzzysort.mjs` files are extremely similar, but that seemed better than the confusion of knowing which file in the root to edit and having rollup potentially overwrite your changes. I've also kept the build files checked in, to keep things simple. A nice future improvement I'm willing to contribute would be github actions CI and creating a PR when the files drift?

hugo-vrijswijk committed:
Add ESM and rollup build

Fixes #94

Adds a rollup config file to build ESM, CJS, and UMD bundles. Based on the source in src/fuzzysort.js.

Checks in the new esm file and the updated cjs and umd files.

Also updates the package.json to include `module` and `browser` fields for the ESM and UMD bundles.

blq commented:
IMHO you have added a bit too much new pretty-print formatting, making the diff larger than it has to be. Can you reduce that a bit to zone in on just the top-level?

hugo-vrijswijk commented:
> IMHO you have added a bit too much new pretty-print formatting, making the diff larger than it has to be. Can you reduce that a bit to zone in on just the top-level?

Yeah, sure. I admit some of it was my editor auto-formatting by itself. What makes the diff difficult is that the source file is now `src/fuzzysort.js` instead of `fuzzysort.js`. So the easiest to diff is probably by comparing `src/fuzzysort.js` with an older version from clipboard or something.

I've pushed a new version that minimizes formatting changes. The diff is still weird, but that is due to the new file not being compared with the old source. Comparing `fuzzysort.js` from master and `src/fuzzysort.js` on this branch is the easiest

blq commented on rollup.config.mjs:
```diff
@@ -0,0 +1,23 @@
+export default defineConfig([
+  {
+    input: 'src/fuzzysort.js',
+    output: { banner, exports, file: pkg.browser, format: 'umd', name: 'fuzzysort' },
```
while you're at it I suggest also enabling source map output to help debugging.
https://rollupjs.org/configuration-options/#output-sourcemap

hugo-vrijswijk replied:
For just the minified file? Or all? The sourcemap is quite big to be checked in IMO

blq replied:
I don't mean check in, but generate in build. At least there's no source map comment at the end of the fuzzsort.min.js file.

hugo-vrijswijk replied:
But the build output is checked in now. I didn't want to change that compared to how it is now. If I enable sourcemaps a .map file would have to be checked in for each file, since the output is also checked in

blq replied:
mm, maybe a compromise is to add flag to generate .map files but don't check them in? With reasonable assumption users will run a real build themselves. 
And change back to the non-minified version in the test.html for the quick tests.

hugo-vrijswijk replied:
Done. I've enabled sourcemaps and added the maps to the gitignore. Adding the non-minified version for the tests is a bit tricky since the non-minified versions are ESM and CJS, not IIFE format. Debugging does work great with sourcemaps and the minified files now

blq commented:
> > IMHO you have added a bit too much new pretty-print formatting, making the diff larger than it has to be. Can you reduce that a bit to zone in on just the top-level?
> 
> Yeah, sure. I admit some of it was my editor auto-formatting by itself. What makes the diff difficult is that the source file is now `src/fuzzysort.js` instead of `fuzzysort.js`. So the easiest to diff is probably by comparing `src/fuzzysort.js` with an older version from clipboard or something.
> 
> I've pushed a new version that minimizes formatting changes. The diff is still weird, but that is due to the new file not being compared with the old source. Comparing `fuzzysort.js` from master and `src/fuzzysort.js` on this branch is the easiest

When comparing to the /src version it's much better now yes. 
Small thing I noticed is the github link and version nr at the top comment is missing.

hugo-vrijswijk commented:
Those are added to the build output. That way the version number is always up-to-date

blq commented:
> Those are added to the build output. That way the version number is always up-to-date

ah, ok

hugo-vrijswijk commented:
@farzher could you take a look? 🙏

hugo-vrijswijk commented:
@farzher gentle reminder. I'd appreciate if you could take a look 😊

hugo-vrijswijk committed:
Use TSDown and remove UMD build

Only build esm and cjs formats using TSDown. Remove UMD build. Also use ESM import in test files.

hugo-vrijswijk commented:
I've updated this PR to only include a ESM and CJS build. The built files are no longer checked in. Even adding CJS is debatable at this point, but it's easy to add. Both versions are built using [tsdown](https://tsdown.dev/). Any users of the UMD build can easily use the ESM version, as that is the standard for modules on both the web and node. I've updated the tests to use the esm version.

For anyone looking at this, I have published this separately as https://www.npmjs.com/package/fuzziersort. Which should be a drop-in replacement until this gets merged (if it does at all).

cooperwalbrun commented:
> I've updated this PR to only include a ESM and CJS build. The built files are no longer checked in. Even adding CJS is debatable at this point, but it's easy to add. Both versions are built using [tsdown](https://tsdown.dev/). Any users of the UMD build can easily use the ESM version, as that is the standard for modules on both the web and node. I've updated the tests to use the esm version.
> 
> For anyone looking at this, I have published this separately as https://www.npmjs.com/package/fuzziersort. Which should be a drop-in replacement until this gets merged (if it does at all).

Thank you for taking the time to patch this. This package had become unusable for my purposes without ESM compatibility. It seems that this project might not be actively maintained anymore, so I think it is probable that more and more people will switch to your fork.
